### PR TITLE
Optimize `MutList.reserve!`

### LINKED ISSUE #2357: Optimize MutList.reserve!
- [ ] Reimplement `MutList.reserve!` using `Array.copyOfRange` (depends on #2358)

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #2357

### CODE REVIEW

**Comment by jaschdoc** [comment]
Do you reckon `MutDeque.copyElements!` is now slower than before?

**Comment by jaschdoc** [comment]
Hmm... apparently bitwise-and `&&&` is 2-3 times slower than using `mod`... how come?

File: main/src/library/MutDeque.flix [hunk 1]
```diff
@@ -185,28 +181,38 @@ namespace MutDeque {
...
 
     ///
     /// Copies the elements from `a` to `a1`. Mutates the array `a1`.
     ///
-    def copyElements!(f: Int32, b: Int32, a: Array[a], a1: Array[a]): Unit & Impure =
-        let c = Array.length(a);
-        if (f < b) { // If this predicate is true the elements do not "wrap around" in the array, i.e. the elements are laid out sequentially from [0 .. b].
-            Array.updateSequence!(0,     a[f .. b], a1)
+    def copyElements!(f: Ref[Int32], b: Ref[Int32], a: Ref[Array[a]], c1: Int32): Unit & Impure =
+        let f1 = deref f;
```
Comment (magnus-madsen) [inline main/src/library/MutDeque.flix:195]: Why so complicated? 

Assume we have from `begin` to `end`. Assume the length is `l` and we want to double to `2 x l`.

Could we not just `copyRange(begin, 2 x l, arr`)?

The deque would then start from `0` and go to `end - begin` or something like that.

**Comment by magnus-madsen** [comment]
The MutList changes are easy to understand. What the hell happened with `MutDeque`? :D

**Comment by jaschdoc** [comment]
Haha yeah that's true. `MutDeque` has some complicated logic for deciding how to expand the array so it's not as easy as just replacing with `copyOfRange`... I think?

**Comment by magnus-madsen** [comment]
Sorry for letting this hang. Can we split into two. I'd like to merge in the `MutList` part.

**Comment by jaschdoc** [comment]
Ready to merge

**Pull request merged**